### DISCUSSION TITLE
Modified stack layout (Help please!)

### DISCUSSION THREAD

**Discussion by Barbaross93** [qtile/qtile#2136] [Q&A]
I'm trying to create a custom Stack layout where instead of splitting the windows evenly within a column, I want to be able to open side columns with little preview windows similar to the zoomy layout for each stack. The focused window (or last focused in the stack) will effectively still be in the main column, but the sides of the screen create space to allow small preview windows for other windows in the stack. For simplicity, I'm trying to set it up to for only two columns with the previews on the outer left and right. In the future, if this gets enough traction, maybe modifying it in such a way that for odd numbers of stacks, the preview should go above/below the main column and even numbers of stacks should have the previews on the left or right side of the main columns depending on which side the columns are with respect to the screen/monitor. For now, I'm focused on getting the `configure` function to be how it should be. This is the monstrosity I have pieced together so far by taking code from the zoomy and stack layouts:
```
def configure(self, client, screen_rect):
        for i, s in enumerate(self.stacks):
            if client in s:
                break
        else:
            client.hide()
            return

        if client.has_focus:
            px = self.border_focus
        else:
            px = self.border_normal

        column_width = int(screen_rect.width / len(self.stacks))
        xoffset = screen_rect.x + i * column_width
        window_width = column_width - 2 * self.border_width
        half_screen = int(screen_rect.width / 2)
        f_left, left = screen_rect.hsplit(self.columnwidth - half_screen)
        right, f_right = screen_rect.hsplit(half_screen - self.columnwidth)

        if s.split:
            if client is self.clients.current_client:
                if client == s.cw:
                    client.place(
                        xoffset,
                        screen_rect.y,
                        window_width,
                        screen_rect.height - 2 * self.border_width,
                        self.border_width,
                        px,
                        margin=self.margin,
                    )
                    client.unhide()
                else:
                    client.hide()
            else:
                h = f_left.width * left.height // left.width
                client_index = self.clients.index(client)
                focused_index = self.clients.current_index
                offset = client_index - focused_index - 1
                if offset < 0:
                    offset += len(self.clients)
                    if h * (len(self.clients) - 1) < f_left.height:
                        client.place(
                            f_left.x,
                            f_left.y + h * offset,
                            f_left.width,
                            h,
                            self.border_width,
                            px,
                            margin=self.margin,
                        )
                    else:
                        hh = (left.height - h) // (len(self.clients) - 1)
                        client.place(
                            left.x,
                            left.y + hh * offset,
                            left.width,
                            h,
                            0,
                            None,
                            margin=self.margin,
                        )

                j = f_right.width * right.height // right.width
                client_index = self.clients.index(client)
                focused_index = self.clients.current_index
                offset = client_index - focused_index - 1
                if offset < 0:
                    offset += len(self.clients)
                    if j * (len(self.clients) - 1) < f_right.height:
                        client.place(
                            f_right.x,
                            f_right.y + h * offset,
                            f_right.width,
                            j,
                            self.border_width,
                            px,
                            margin=self.margin,
                        )
                    else:
                        jj = (right.height - h) // (len(self.clients) - 1)
                        client.place(
                            right.x,
                            right.y + hh * offset,
                            right.width,
                            j,
                            0,
                            None,
                            margin=self.margin,
                        )
            client.unhide()
        else:
            if client == s.cw:
                client.place(
                    xoffset,
                    screen_rect.y,
                    window_width,
                    screen_rect.height - 2 * self.border_width,
                    self.border_width,
                    px,
                    margin=self.margin,
                )
                client.unhide()
            else:
                client.hide()
```
This 100% will not work. Since this is my first time doing anything serious in python, I would genuinely appreciate some guidance with this

**Comment by elParaguayo** [comment]
If you haven't played with layouts before (and I haven't), I'd suggest making very small changes to existing code and seeing how the layout is impacted. Play with one particular function and see how it behaves and then build up from there.

When I'm testing code, my workflow tends to look like this:
1. create a new branch 
2. modify code
2. start a new X session
3. launch qtile 
4. look at the logs to see why qtile crashed!

I'm intrigued with your idea so I'm tempted to take a look when I have some time but there are some other things I need to finish first.

**Comment by Barbaross93** [comment]
Thanks @elParaguayo! You're right, I was hasty to get the overall layout figured out, but I should really stick to babysteps. I really should get better with git, still haven't figured out how to make separate branches.

I'm glad I'm not alone with an interest in this idea! I wasn't sure if I was clear in describing it; I tried to "draw" it with characters but it turned out terribly.

As an aside, I wonder how difficult it would be to instead of using small preview windows, to draw a small container with the app icon centered?

**Comment by elParaguayo** [comment]
Just been looking at this. The way I wanted to do this won't work.

I was thinking about adding a new, unmanaged window to the layout and then drawing the icon of the clients to that window. The window would be shown when changing clients but I also wanted to have a timer to hide it after a certain period.

Both of these need the current qtile object to be available to the layout but it's not currently passed (which is fair enough, there's no real need for layouts to have access to this).

EDIT: Well, I can access `qtile` in a hacky way. Will keep playing with this...

**Comment by elParaguayo** [comment]
OK. So, I can get icons of windows in stack and display them. Need to add a border for the current window so you can see where you are in the stack. Not a bad start though...

![stack](https://user-images.githubusercontent.com/946265/105612789-0b4da100-5db6-11eb-8053-8fb9676663eb.png)

**Comment by Barbaross93** [comment]
That's amazing!!! Could you share what you have so far?

**Comment by elParaguayo** [comment]
Just playing around with it some more. Will upload something to my fork later tonight.

Lots of hard coded values at the moment!

**Comment by elParaguayo** [comment]
This is where I'm at now:

https://user-images.githubusercontent.com/946265/105638600-5aa1d900-5e6b-11eb-8152-7586b0f206fc.mp4


It's on this branch of my fork: https://github.com/elParaguayo/qtile/tree/layout_stack_display

Lots of things that still need to be worked on:

- Hide icons after timeout
- Better handling of icon size
- Better handling of drawing border
- etc. !

**Comment by Barbaross93** [comment]
Thank you! When I get a chance, I'll poke at this as well! That is an interesting idea though, to have the icons disappear after a certain amount of time. I was thinking of creating a function to toggle their visibility, and then add a keybinding to do that

**Comment by elParaguayo** [comment]
Binding to a key press is easy enough. However, the most important bit outstanding which I forgot to mention is that we need to hide/destroy the preview windows when the layout is changed. If you don't, they will stay where they are.
